src/Eshopworld.WorkerProcess/LeaseAllocator.cs

```diff
@@ -34,83 +35,103 @@ public class LeaseAllocator : ILeaseAllocator
         /// <inheritdoc />
         public async Task<ILease> AllocateLeaseAsync(Guid instanceId)
         {
+            //Lease allocation following design documented here : https://eshopworld.atlassian.net/wiki/spaces/EPLT/pages/2345435290/Worker+Process
+            _telemetry.Publish(new LeaseAcquisitionEvent(instanceId, _options.Value.WorkerType,
+                _options.Value.Priority, $"Lease allocation started for instance {instanceId}"));
+           
             var lease = await _leaseStore.ReadByLeaseTypeAsync(_options.Value.WorkerType).ConfigureAwait(false);
+            if (lease?.LeasedUntil != null && lease.LeasedUntil.Value > ServerDateTime.UtcNow)
+                return null;
+            
+            await _leaseStore.AddLeaseRequestAsync(_options.Value.WorkerType,_options.Value.Priority,instanceId).ConfigureAwait(false);
 
-            if (lease == null)
-            {
-                var leaseResult = await _leaseStore.TryCreateLeaseAsync(_options.Value.WorkerType,
-                    _options.Value.Priority, instanceId).ConfigureAwait(false);
+            // backoff to allow other workers to add their lease request
+            var delay = _allocationDelay.Calculate(_options.Value.Priority, _options.Value.LeaseInterval);
+            await Task.Delay(delay);
 
-                if (leaseResult.Result)
-                {
-                    lease = leaseResult.Lease;
-                }
-                else
-                {
-                    // another worker lease created the lease before this worker read the created lease
-                    lease = await _leaseStore.ReadByLeaseTypeAsync(_options.Value.WorkerType).ConfigureAwait(false);
-                }
-            }
+            _telemetry.Publish(new LeaseAcquisitionEvent(instanceId, _options.Value.WorkerType,
+                _options.Value.Priority, $"Leader election starting"));
 
-            // check lease priority is less than current worker priority and lease is not acquired
-            if (lease.Priority <= _options.Value.Priority && lease.LeasedUntil == null ||
-                lease.LeasedUntil.HasValue && lease.LeasedUntil.Value < ServerDateTime.UtcNow)
-            {
-                lease.Priority = _options.Value.Priority;
-                lease.InstanceId = instanceId;
+            var winnerInstanceId = await _leaseStore.SelectWinnerRequestAsync(_options.Value.WorkerType);
 
-                var updateResult = await _leaseStore.TryUpdateLeaseAsync(lease).ConfigureAwait(false);
+            if (!winnerInstanceId.HasValue || winnerInstanceId != instanceId)
+            {
+                var reason = !winnerInstanceId.HasValue
+                    ? "There are no LeaseRequests to evaluate"
```

Comment:
This is rather strange case and should not happen - I would almost consider this to be an exception

Reply:
Agreed to record as a different telemetry event and to consider an alert

Reply:
added a new WinnerNotExistingEvent to handle this scenario